eds: Adding eds caching support to grpc-mux

adisuissa opened the pull request:
Commit Message: eds: Adding eds caching support to grpc-mux
Additional Description:

Continuation of PR #28079 (as part of the work for issue #26749).
Currently after an EDS-cluster update, Envoy waits for an EDS response. If a timeout occurs, the EDS-cluster will be used without endpoints.
This PR adds the use of caching into the GrpcMux. The GrpcMux object adds an EDS resource to the cache when it is received/updated, and removes it when there are no longer subscriptions (watchers).
A runtime flag is added to disable the use of the cache, and will be enabled in a future PR when ADS is used.

Next PR will plumb this into ADS, and add fetching of resources from the cache as part of the EdsClusterImpl.

The entire change can be looked here: https://github.com/adisuissa/envoy/commit/f0b7ac8ab85776947a8332efe1c3759e8714b9ed

Risk Level: Low - the disabled runtime flag should prevent the use of the cache in non-tests code.
Testing: Added unit tests.
Docs Changes: N/A.
Release Notes: N/A (future PR).
Platform Specific Features: N/A.
Runtime guard: disabled by default: `envoy_restart_features_use_eds_cache_for_ads`

adisuissa commented:
Assigning @htuch for a high-level review.
/assign @htuch

htuch commented:
@adisuissa before diving in to review, can you remind me what the plan is for the scenario when we have a cluster being upgraded from non-TLS to TLS? Do we still have an opportunity to wait for the appropriate endpoints, or do we want to discourage the use of synchronization here?

adisuissa commented:
> @adisuissa before diving in to review, can you remind me what the plan is for the scenario when we have a cluster being upgraded from non-TLS to TLS? Do we still have an opportunity to wait for the appropriate endpoints, or do we want to discourage the use of synchronization here?

Yeah, we still support the upgrade from non-TLS to TLS.

Envoy will wait for the EDS assignment, and only if it isn't received (timeout), it will be fetched from the cache.
The main benefit - not breaking anything, because previously an empty assignment will be set on timeout, making the cluster unusable.
The downside is that there is a delay between receiving the cluster update until that cluster is actually used (with the old endpoints). It is the same behavior as before, but may be a bit non-intuitive from operations point-of-view.

htuch commented on source/extensions/config_subscription/grpc/grpc_mux_impl.cc:
```diff
@@ -412,6 +419,19 @@ void GrpcMuxImpl::processDiscoveryResources(const std::vector<DecodedResourcePtr
     // updates in the message for EDS/RDS.
     if (!found_resources.empty()) {
       watch->callbacks_.onConfigUpdate(found_resources, version_info);
+      // Resource cache is only used for EDS resources.
```
As discussed offline, it would be good to document somewhere (maybe an issue) the strategy we're taking around new / old gRPC mux impl. It's totally fine to do this particular PR for both based on the plan you have, but it would be good to make this visible to others working in this space what is going on and telegraph the old / new gRPC mux behavioral change and migration story.

htuch commented on source/extensions/config_subscription/grpc/grpc_mux_impl.h:
```diff
@@ -99,17 +105,29 @@ class GrpcMuxImpl : public GrpcMux,
+        if (eds_resources_cache_.has_value()) {
+          // All resources are to be removed, so remove them from the cache.
+          for (const auto& resource_name : resources_) {
+            eds_resources_cache_->removeResource(resource_name);
```
How come the cache does not follow a RAII model?

htuch commented on source/extensions/config_subscription/grpc/grpc_mux_impl.h:
```diff
@@ -131,7 +149,12 @@ class GrpcMuxImpl : public GrpcMux,
+      // set until the end the function and computing the diff.
+      // Temporarily keep the resources prior to the update to find which ones
+      // were removed.
+      std::set<std::string> previous_resources;
```
Do we do a similar pattern elsewhere (IIRC) for mapping SotW to delta and back? Should we have some smarter resource differencer that can be shared?

adisuissa replied:
The delta-xDS/unified-mux use the watch-map [updateWatchInterest](https://github.com/envoyproxy/envoy/blob/1f67c88f26744e1f7c21c0e65798d77d69efeac8/source/extensions/config_subscription/grpc/watch_map.cc#L51) in order to achieve this.
Refactoring it is essentially what the unified-mux has done, as the non-unified SotW doesn't use a watch-map.

htuch commented on source/extensions/config_subscription/grpc/xds_mux/grpc_mux_impl.cc:
```diff
@@ -1,5 +1,6 @@
 #include "source/extensions/config_subscription/grpc/xds_mux/grpc_mux_impl.h"
 
+#include "envoy/config/endpoint/v3/endpoint.pb.h"
```
This is the 3rd repetition of this logic, how come we have 3 mux implementations again? Was this the one old, one new, one unified? This seems really bad situation to do just about anything in xDS space :/

alyssawilk commented:
/wait on review comments

htuch commented on source/common/runtime/runtime_features.cc:
```diff
@@ -115,6 +115,9 @@ FALSE_RUNTIME_GUARD(envoy_reloadable_features_always_use_v6);
 FALSE_RUNTIME_GUARD(envoy_reloadable_features_refresh_rtt_after_request);
+// TODO(adip) this will be enabled by default once the work on the feature is
+// done in EdsClusterImpl.
+FALSE_RUNTIME_GUARD(envoy_restart_features_use_eds_cache_for_ads);
```
Please add to release notes coverage of this feature and runtime guard.

adisuissa replied:
The release note will be added once the feature is fully integrated into EDS (next PR will include the changes needed in the EDS subscription - see [the entire feature implementation](https://github.com/adisuissa/envoy/commit/f0b7ac8ab85776947a8332efe1c3759e8714b9ed)).

htuch commented on test/extensions/config_subscription/grpc/delta_subscription_impl_test.cc:
```diff
@@ -159,13 +159,15 @@ TEST_P(DeltaSubscriptionNoGrpcStreamTest, NoGrpcStream) {
         *stats_store.rootScope(), rate_limit_settings, local_info, false,
         std::make_unique<NiceMock<MockCustomConfigValidators>>(), std::move(backoff_strategy),
-        /*xds_config_tracker=*/XdsConfigTrackerOptRef());
+        /*xds_config_tracker=*/XdsConfigTrackerOptRef(),
```
I think given the growth here of names params it might make ense to switch to an options struct. WDYT?

adisuissa replied:
Agreed, this will be ~800 lines change. As discussed offline, we'll defer it to a followup PR.

htuch commented on test/extensions/config_subscription/grpc/new_grpc_mux_impl_test.cc:
```diff
@@ -561,6 +565,182 @@ TEST_P(NewGrpcMuxImplTest, Shutdown) {
 }
 
+// Validate that an EDS resource is cached if there's a cache.
+TEST_P(NewGrpcMuxImplTest, CacheEdsResource) {
```
Do we have any integration tests?

adisuissa replied:
GrpcMux doesn't have integration tests. There will be after the feature is fully implemented in the next PR (see [the entire feature implementation](https://github.com/adisuissa/envoy/commit/f0b7ac8ab85776947a8332efe1c3759e8714b9ed)).

htuch reviewed (changes requested):
LGTM implementation modulo nits. Have you verified no regression in test coverage?

adisuissa commented:
> LGTM implementation modulo nits. Have you verified no regression in test coverage?

That's a good comment. I've added more tests to cover the call to the cache getter (it was only covered by the integration test that was planned in the followup PR).
Also added some tests to the unified_mux implementation, to make it on par with the non-unified mux.

htuch reviewed (dismissed):
LGTM, thanks!

adisuissa commented:
/retest
